### PR TITLE
Reintroduce team switch kit

### LINKED ISSUE #732: /team force cannot bypass blitz
Previously, PGM would allow `/team force` to work during a Blitz match. However, it no longer workers.

### PR DESCRIPTION
**Pull request by applenick**
# Reintroduce Team Switch Kit

Quick PR to reimplement the team switch kit back into PGM. Also funny enough this fixes #732 

Thanks to @CrazyisCreeps for helping me test 😄 

A little background, this kit is primarily used for the Infection gamemode. But also could have some nifty applications when applied to new maps 😁  

## Example XML:
```xml
 <kit id="switch-to-apple-team" force="true">
     <team-switch team="apples" title="true"/>
 </kit>
```

## Video
Updated

https://user-images.githubusercontent.com/3377659/107834175-5ff79280-6d4a-11eb-8a8a-489bf0d86a22.mp4

Original Video

https://user-images.githubusercontent.com/3377659/107683199-25143280-6c56-11eb-8d2d-bd1c71e4fe2f.mp4

Signed-off-by: applenick <applenick@users.noreply.github.com>

### CODE REVIEW

**Comment by CoWinkKeyDinkInc** [comment]
That's pretty neat, what sort of things could/has this been used for? :D

**Comment by RuedigerLP** [comment]
> That's pretty neat, what sort of things could/has this been used for? :D

Infection gamemode

File: core/src/main/java/tc/oc/pgm/kits/TeamSwitchKit.java [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+
+  private final TeamFactory team;
+
+  public TeamSwitchKit(TeamFactory team) {
+    this.team = team;
+  }
+
+  @Override
+  public void applyDelayed(MatchPlayer player, boolean force) {
+    TeamMatchModule tmm = player.getMatch().getModule(TeamMatchModule.class);
```
Comment (Pablete1234) [inline core/src/main/java/tc/oc/pgm/kits/TeamSwitchKit.java]: i'd just require team match module, there's no point in silently ignoring when someone tries to use this in FFA

**Comment by CrazyisCreeps** [comment]
> That's pretty neat, what sort of things could/has this been used for? :D

I made a freeze tag styled map with this module back on Stratus where when tagged you were swapped to a "Frozen Team" that basically prevented you from moving. I think it used a combination of attacker and victim kits as well to really make it work.

**Comment by applenick** [comment]
> > That's pretty neat, what sort of things could/has this been used for? :D
> 
> I made a freeze tag styled map with this module back on Stratus where when tagged you were swapped to a "Frozen Team" that basically prevented you from moving. I think it used a combination of attacker and victim kits as well to really make it work.

Woah that sounds cool! Would be interested to see what other cool map concepts we can implement with this too 😄

**Comment by Electroid** [comment]
Could we show the `You joined X` message as a sub-title? We want to make it obvious that they have switched teams

**Comment by Pablete1234** [comment]
> Could we show the `You joined X` message as a sub-title? We want to make it obvious that they have switched teams

Would be nice, however, i'd keep an XML tag to be able to turn it off

File: core/src/main/java/tc/oc/pgm/kits/TeamSwitchKit.java (hunk 1 above)
Comment (applenick) [reply core/src/main/java/tc/oc/pgm/kits/TeamSwitchKit.java]: Great idea! Will adjust 👍

**Comment by applenick** [comment]
@Electroid sure thing! That's an excellent idea, may have to change how the join messages are sent but should be manageable, I'll work on this in a little bit.

@Pablete1234 also will include a tag to choose

**Comment by applenick** [comment]
@Electroid & @Pablete1234 

Example of two different regions applying the team switch kit
One with `title="true"` the other with `title="false"`

https://user-images.githubusercontent.com/3377659/107834175-5ff79280-6d4a-11eb-8a8a-489bf0d86a22.mp4

File: core/src/main/java/tc/oc/pgm/kits/KitParser.java [hunk 2]
```diff
@@ -670,4 +673,17 @@ public ShieldKit parseShieldKit(Element parent) throws InvalidXMLException {
         XMLUtils.parseDuration(el.getAttribute("delay"), ShieldParameters.DEFAULT_DELAY);
     return new ShieldKit(new ShieldParameters(health, rechargeDelay));
   }
+
+  public TeamSwitchKit parseTeamSwitchKit(Element parent) throws InvalidXMLException {
+    Element el = XMLUtils.getUniqueChild(parent, "team-switch");
+    if (el == null) return null;
+
+    boolean showTitle = XMLUtils.parseBoolean(el.getAttribute("title"), true);
```
Comment (Electroid) [inline core/src/main/java/tc/oc/pgm/kits/KitParser.java]: `show-title`

**Review by Electroid — DISMISSED** [review]
Just 1 change, otherwise looks good.

**Comment by applenick** [comment]
@Electroid all good now 👍

**Review by Pablete1234 — DISMISSED** [review]

File: core/src/main/java/tc/oc/pgm/teams/TeamMatchModule.java [hunk 3]
```diff
@@ -573,9 +612,32 @@ public boolean kickPlayerOffTeam(Team kickFrom, boolean forBalance) {
...
+      Component joinMsg = translatable("join.ok.team", event.getNewParty().getName());
+      boolean title = false;
+
+      // Players with team switch kit, check for title value and remove from map
+      if (hasTeamSwitchKit(player)) {
+        title = showTeamSwitchTitle(player);
+        setTeamSwitchKit(player, false, false);
+      }
+
+      if (title) {
```
Comment (Pablete1234) [inline core/src/main/java/tc/oc/pgm/teams/TeamMatchModule.java]: 1.7 players can't see titles, they need to keep the chat messages

Comment (applenick) [reply core/src/main/java/tc/oc/pgm/teams/TeamMatchModule.java]: Oh good catch, I'll add in a legacy check

**Review by Pablete1234 — APPROVED** [review]

**Review by Electroid — APPROVED** [review]

**Pull request merged**